frontend/src/app/pages/DashBoardPage/hooks/useWidgetAutoFetch.ts

```diff
@@ -21,6 +21,7 @@ import { BoardInfoContext } from '../components/BoardProvider/BoardInfoProvider'
 import { VizRenderMode, Widget } from '../pages/Board/slice/types';
 import { isElView } from '../utils/board';
 
+let timer: NodeJS.Timeout;
```

Comment:
I think it will be better

```diff

let timer: NodeJS.Timeout | null;

+function clearTimer() {
+  if (timer) {
+    clearInterval(timer);
+    timer = null;
+  }
+}

export default function useWidgetAutoFetch(
  widget: Widget,
  renderMode: VizRenderMode,
  rectRef: React.RefObject<HTMLDivElement>,
  rendered: boolean,
) {
  const { visible: boardVisible } = useContext(BoardInfoContext);
  const { onWidgetGetData } = useContext(WidgetActionContext);

  useEffect(() => {
    if (
      rendered &&
      boardVisible &&
      widget.config.frequency > 0 &&
      widget.config.autoUpdate
    ) {
+     clearTimer();
      timer = setInterval(() => {
        const elShow = isElView(rectRef.current);
        if (elShow) {
          onWidgetGetData(widget);
        }
      }, +widget.config.frequency * 1000);
    }
+   return clearTimer;
  }, [
    boardVisible,
    widget,
    widget.config.autoUpdate,
    onWidgetGetData,
    renderMode,
    rendered,
    rectRef,
  ]);
}

```

Reply:
maybe

Reply:
but I already wrote return clearTimer in useEffect hook, why do I need to clearTimer() before the set setInterval @Zclhlmgqzc

Reply:
> but I already wrote return clearTimer in useEffect hook, why do I need to clearTimer() before the set setInterval @Zclhlmgqzc 

Because you've changed its scope

Reply:
说实话我不明白为什么要把 timer 弄成一个单例？
如果是业务需要 那么需要我上面说的那样如果重新赋值前要做个清除

如果是现在这样 会有问题 仪表板 定义两个图表分别设置自动刷新 10s、15s 切换别的仪表板 会发现 只清除了一个定时任务 

看了看逻辑 不能用单例 还是得回滚代码 即在 useEffect 里定义 timer 

![image](https://user-images.githubusercontent.com/31752986/163586653-26e2ae4d-47cf-43c9-b2a8-cdcfea57c50e.png)

![image](https://user-images.githubusercontent.com/31752986/163586600-c20fd003-7ea7-4c17-b895-a5e6143af181.png)

Reply:
我发现确实有问题，很感谢你的提醒。useEffect 里定义 timer 应该是更好的选择。我当时没有考虑到这一点。我会尽快修改一下这里。再次感谢你的提醒👍🥂

Reply:
new pr  https://github.com/running-elephant/datart/pull/1207  这样应该比较合适了，right?